crates/ruff_db/src/system.rs

```diff
@@ -130,20 +139,20 @@ impl FileType {
 #[derive(Debug)]
 pub struct DirectoryEntry {
     path: SystemPathBuf,
-    file_type: Result<FileType>,
+    file_type: FileType,
 }
 
 impl DirectoryEntry {
-    pub fn new(path: SystemPathBuf, file_type: Result<FileType>) -> Self {
+    pub fn new(path: SystemPathBuf, file_type: FileType) -> Self {
         Self { path, file_type }
     }
 
     pub fn path(&self) -> &SystemPath {
         &self.path
     }
 
-    pub fn file_type(&self) -> &Result<FileType> {
-        &self.file_type
+    pub fn file_type(&self) -> FileType {
+        self.file_type
     }
 }
```

Comment:
Heh, doesn't this just reverse the change you asked me to make in https://github.com/astral-sh/ruff/pull/12289#discussion_r1674657856? ;)

Reply:
Yeah, that's probably true. I ran into the issue that `io::Error` isn't `Clone`. I could have worked around this problem by destructing both the `path` and the `result` from `file_type`, but I'm unsure if it's worth it. 

The main advantage of `std::fs::DirEntry::file_type` is that the operation is lazy. This is different in our implementation, where we call `file_type` when returning the `DirEntry` (so it is eager). 

Now, there's a semantic difference in the new implementation. The `read_dir` call will fail when calling `file_type` fails. I don't know in what situations it is possible to know that the entry is there, but it fails to read the file type. 

I just spent some time looking into what `walker` does. It seems `walker' also eagerly reads the file type and returns an `Err` if it fails to retrieve it. 

Ideally, I think we would make the `file_type` call lazy, but that's quiet a bit more complicated because it would require holding on to `System`. 

TLDR: Not sure, but it felt like the most straightforward path and it's unclear to me what the benefit is of keeping the Result.

Reply:
That all makes sense to me. If we're no longer storing a `Result`, though, I think I'd also make the following simplifications:

```diff
diff --git a/crates/ruff_db/src/system.rs b/crates/ruff_db/src/system.rs
index 2805137ec..5b1ce5b12 100644
--- a/crates/ruff_db/src/system.rs
+++ b/crates/ruff_db/src/system.rs
@@ -115,7 +115,7 @@ impl Metadata {
     }
 }
 
-#[derive(Copy, Clone, Eq, PartialEq, Debug, Hash)]
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Hash, PartialOrd, Ord)]
 pub enum FileType {
     File,
     Directory,
@@ -136,7 +136,7 @@ impl FileType {
     }
 }
 
-#[derive(Debug)]
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
 pub struct DirectoryEntry {
     path: SystemPathBuf,
     file_type: FileType,
@@ -155,9 +155,3 @@ impl DirectoryEntry {
         self.file_type
     }
 }
-
-impl PartialEq for DirectoryEntry {
-    fn eq(&self, other: &Self) -> bool {
-        self.path == other.path
-    }
-}
diff --git a/crates/ruff_db/src/system/os.rs b/crates/ruff_db/src/system/os.rs
index 2b7dd397d..99fff9449 100644
--- a/crates/ruff_db/src/system/os.rs
+++ b/crates/ruff_db/src/system/os.rs
@@ -283,7 +283,7 @@ mod tests {
             .unwrap()
             .map(Result::unwrap)
             .collect();
-        sorted_contents.sort_by(|a, b| a.path.cmp(&b.path));
+        sorted_contents.sort();
 
         let expected_contents = vec![
             DirectoryEntry::new(tempdir_path.join("a/bar.py"), FileType::File),
```

Reply:
Oh nice. I'll do that. But I'm not sure about `PartialOrd` on `FileType`. Not sure what the use case (or even the order) is.

Reply:
> Not sure what the use case (or even the order) is.

I guess the use case is that you can easily derive `PartialOrd` on `DirectoryEntry`... but yeah, agreed that that one is questionable :)